types/decimal.go

```diff
@@ -0,0 +1,351 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+// NOTE: never use new(Dec) or else we will panic unmarshalling into the
+// nil embedded big.Int
+type Dec struct {
+	*big.Int `json:"int"`
+}
+
+// number of decimal places
+const Precision = 10
+
+var (
+	precisionExpReuse    = big.NewInt(Precision)
+	ten                  = big.NewInt(10)
+	precisionReuse       = new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
+	precisionMultipliers []*big.Int
+	zeroInt              = big.NewInt(0)
+	tenInt               = big.NewInt(10)
+)
+
+// Set precision multipliers
+func init() {
+	precisionMultipliers = make([]*big.Int, Precision+1)
+	for i := 0; i <= Precision; i++ {
+		precisionMultipliers[i] = calcPrecisionMultiplier(int64(i))
+	}
+}
+
+func precisionInt() *big.Int {
+	return new(big.Int).Set(precisionReuse)
+}
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{new(big.Int).Set(zeroInt)} }
+func OneDec() Dec  { return Dec{precisionInt()} }
+
+// calculate the precision multiplier
+func calcPrecisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
+	}
+	zerosToAdd := Precision - prec
+	multiplier := new(big.Int).Exp(tenInt, big.NewInt(zerosToAdd), nil)
+	return multiplier
+}
+
+// get the precision multiplier. Do not mutate result.
+func precisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
+	}
+	return precisionMultipliers[prec]
+}
+
+// create a new Dec from integer assuming whole numbers
+// CONTRACT: prec <= Precision
+func NewDec(i, prec int64) Dec {
```

Comment:
Is there a reason why we're requiring precision this time, compared to the approach with rationals which was to assume denominator=1 if not specified. Seems like we can do the same here with default precision=0

Reply:
I agree. We should make the default param prec is 0, and then a second function with newDecWithPrec

Reply:
meh - okay

Reply:
k added in